typing/typecore.ml

```diff
@@ -3199,95 +3226,137 @@ let rec approx_type env sty =
+let rec type_approx env sexp ty_expected =
   let loc = sexp.pexp_loc in
   match sexp.pexp_desc with
-    Pexp_let (_, _, e) -> type_approx env e
+    Pexp_let (_, _, e) -> type_approx env e ty_expected
   | Pexp_function (params, c, body) ->
-      type_approx_function env params c body ~loc
-  | Pexp_match (_, {pc_rhs=e}::_) -> type_approx env e
-  | Pexp_try (e, _) -> type_approx env e
-  | Pexp_tuple l ->
-    let labeled_tys = List.map (fun (label, _) -> label, newvar ()) l in
-    newty (Ttuple labeled_tys)
-  | Pexp_ifthenelse (_,e,_) -> type_approx env e
-  | Pexp_sequence (_,e) -> type_approx env e
+      type_approx_function env params c body ty_expected ~loc
+  | Pexp_match (_, {pc_rhs=e}::_) -> type_approx env e ty_expected
+  | Pexp_try (e, _) -> type_approx env e ty_expected
+  | Pexp_tuple l -> type_tuple_approx env sexp.pexp_loc ty_expected l
+  | Pexp_ifthenelse (_,e,_) -> type_approx env e ty_expected
+  | Pexp_sequence (_,e) -> type_approx env e ty_expected
   | Pexp_constraint (e, sty) ->
-      let ty = type_approx env e in
-      type_approx_constraint env ty (Pconstraint sty) ~loc
-  | Pexp_coerce (e, sty1, sty2) ->
-      let ty = type_approx env e in
-      type_approx_constraint env ty (Pcoerce (sty1, sty2)) ~loc
+      let ty_expected =
+        type_approx_constraint env (Pconstraint sty) ~loc ty_expected
+      in
+      type_approx env e ty_expected
+  | Pexp_coerce (_, sty1, sty2) ->
+      ignore @@
+      type_approx_constraint env (Pcoerce (sty1, sty2)) ~loc ty_expected
   | Pexp_pack (_, Some ptyp) ->
-      let ty = newvar () in
       let sty = Ast_helper.Typ.package ~loc ptyp in
-      type_approx_constraint env ty (Pconstraint sty) ~loc
-  | _ -> newvar ()
+      ignore @@
+      type_approx_constraint env (Pconstraint sty) ~loc ty_expected
+  | _ -> ()
 
-and type_approx_function env params c body ~loc =
-  (* We can approximate types up to the first newtype parameter, whereupon
-     we give up.
-  *)
-  match params with
-  | { pparam_desc = Pparam_val (label, default, pat) } :: params ->
-      type_approx_fun env label default pat
-        (type_approx_function env params c body ~loc)
-  | { pparam_desc = Pparam_newtype _ } :: _ ->
-      newvar ()
-  | [] ->
-    let body_ty =
-      match body with
-      | Pfunction_body body ->
-          type_approx env body
-      | Pfunction_cases ({pc_rhs = e} :: _, _, _) ->
-          newty (Tarrow (Nolabel, newvar (), type_approx env e, commu_ok))
-      | Pfunction_cases ([], _, _) ->
-          newvar ()
-    in
-    type_approx_constraint_opt env body_ty c ~loc
+and type_tuple_approx (env: Env.t) loc ty_expected l =
+  let labeled_tys = List.map (fun (label, _) -> label, newvar ()) l in
+  let ty = newty (Ttuple labeled_tys) in
+  begin try unify env ty ty_expected with Unify err ->
+    raise(Error(loc, env, Expr_type_clash (err, None, None)))
+  end;
+  List.iter2
+    (fun (_, e) (_, ty) -> type_approx env e ty)
+    l labeled_tys
```

Comment:
Seems to change slightly the set of codes accepted by the type checker because we now go inside tuples.